### PR TITLE
[SL-2530] Feature: Edit relationships

### PR DESCRIPTION
**Pull request by freixedelo**
- drag & drop objects from listing as relationship in relationship panel 
- remove objects from relationship listing

* [x] Feature

#### Jira

https://skylarkplatform.atlassian.net/browse/SL-2530

### CODE REVIEW

File: src/components/panel/panelSections/panelRelationships.component.tsx [hunk 1]
```diff
@@ -1,92 +1,196 @@
...
+                      relationshipObject &&
+                      originalRelationshipObjects &&
+                      parseUpdatedRelationshipObjects(
+                        relationshipObject,
+                        updatedRelationshipObjects,
+                        originalRelationshipObjects,
+                      ).uidsToLink;
+
+                    return (
                       <>
```
Comment (james-wallis) [inline src/components/panel/panelSections/panelRelationships.component.tsx:135]: <Fragment key={obj.uid}>

Comment (james-wallis) [reply src/components/panel/panelSections/panelRelationships.component.tsx:135]: `<Fragment key={obj.uid>`

File: src/components/panel/panelSections/panelRelationships.component.tsx [hunk 2]
```diff
@@ -1,92 +1,196 @@
...
+                          {inEditMode && newUids?.includes(obj.uid) && (
+                            <span
+                              className={
+                                "flex h-6 min-w-6 items-center justify-center rounded-full bg-success px-1 pb-0.5 text-center text-white transition-colors"
+                              }
+                            />
+                          )}
+                          <button
+                            disabled={!inEditMode}
+                            data-testid={`panel-object-content-item-${
```
Comment (james-wallis) [inline src/components/panel/panelSections/panelRelationships.component.tsx:147]: not sure we need this data-testid ?

File: src/components/toast/toast.component.tsx [hunk 3]
```diff
@@ -32,17 +33,24 @@ export const Toast = ({
...
         (type === "default" || !type) && "bg-brand-primary",
         type === "info" && "bg-info text-info-content",
         type === "success" && "bg-success text-success-content",
         type === "warning" && "bg-warning text-warning-content",
         type === "error" && "bg-error text-error-content",
       )}
     >
+      <button
+        onClick={() => toastProps?.deleteToast()}
+        className="absolute top-1 right-2"
```
Comment (james-wallis) [inline src/components/toast/toast.component.tsx:50]: try top-2 it looked a bit close to the top

File: src/constants/skylark.ts [hunk 4]
```diff
@@ -49,6 +49,7 @@ export const DISPLAY_NAME_PRIORITY = [
 ];
 
 export const DROPPABLE_ID = "droppable";
+// export const DROPPABLE_RELATIONSHIPS_ID = "droppable-relationships";
```
Comment (james-wallis) [inline src/constants/skylark.ts:52]: remove

File: src/hooks/useGetObjectRelationships.ts [hunk 5]
```diff
@@ -93,11 +93,12 @@ export const useGetObjectRelationships = (
           objects: parsedObjects,
         };
       })
-    : [];
+    : null;
 
   return {
     ...rest,
     data: relationships,
+    relationships: objectOperations?.relationships,
```
Comment (james-wallis) [inline src/hooks/useGetObjectRelationships.ts:101]: keep this if its used

File: src/hooks/useUpdateObjectRelationships.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  const updateObjectRelationshipsMutation =
+    createUpdateObjectRelationshipsMutation(
+      objectOperations,
+      updatedRelationshipObjects,
+      originalRelationshipObjects,
+    );
+
+  const { mutate, ...rest } = useMutation({
+    mutationFn: ({ uid }: { uid: string }) => {
+      return skylarkRequest<GQLSkylarkUpdateObjectContentResponse>(
```
Comment (james-wallis) [inline src/hooks/useUpdateObjectRelationships.tsx:40]: `GQLSkylarkUpdateObjectContentResponse`
This should be specific to the relationship response

File: src/components/panel/panel.component.tsx [hunk 7]
```diff
@@ -122,7 +139,61 @@ export const Panel = ({
...
+            (relationship) =>
+              relationship.relationshipName === droppedObjectRelationshipName,
+          );
+
+        const isAlreadyAdded = !!droppedObjectRelationshipObjects?.objects.find(
+          ({ uid }) => droppedObject.uid === uid,
+        );
+
+        if (isAlreadyAdded) {
+          toast(
```
[the remark this answers is not in the document]
Comment (freixedelo) [reply src/components/panel/panel.component.tsx:160]: cannot return toast inside `useffect`, could use` return;` tho. but i want to run these commands at the end of everything:
```
setEditMode(true);
clearDroppedObject && clearDroppedObject();
```

File: src/components/panel/panelSections/panelRelationships.component.tsx [hunk 8]
```diff
@@ -1,92 +1,196 @@
...
+  setRelationshipObjects,
+  originalRelationshipObjects,
+  inEditMode,
   language,
+  showDropArea,
 }: PanelRelationshipsProps) => {
   const {
-    data: relationships,
+    data: relationshipsData,
+    relationships = [],
```
[the remark this answers is not in the document]
Comment (freixedelo) [reply src/components/panel/panelSections/panelRelationships.component.tsx:49]: `relationships` is the array of relationships to this particular object. i'm setting the default to an empty array when is `null` so don't need to check if is `null` every time.

File: src/hooks/useGetObjectRelationships.ts (hunk 5 above)
Comment (freixedelo) [reply src/hooks/useGetObjectRelationships.ts:101]: it is

File: src/components/panel/panelSections/panelRelationships.component.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (james-wallis) [reply src/components/panel/panelSections/panelRelationships.component.tsx:49]: Fair

File: src/components/panel/panelSections/panelRelationships.component.tsx [hunk 9]
```diff
@@ -1,92 +1,196 @@
...
+
+  const removeRelationshipObject = (removeUid: string, relationship: string) =>
+    updatedRelationshipObjects &&
+    setRelationshipObjects({
+      updatedRelationshipObjects: updatedRelationshipObjects?.map(
+        (currentRelationship) => {
+          const { objects, relationshipName } = currentRelationship;
+          if (relationshipName === relationship) {
+            //filters current object
+            const o = objects.filter((obj) => obj.uid !== removeUid);
```
[the remark this answers is not in the document]
Comment (james-wallis) [reply src/components/panel/panelSections/panelRelationships.component.tsx:72]: prefer `const filteredObjects`

[the remark this answers is not in the document]
Comment (james-wallis) [reply src/components/panel/panelSections/panelRelationships.component.tsx:72]: and remove comment

File: src/components/panel/panelSections/panelRelationships.component.tsx (hunk 1 above)
Comment (james-wallis) [reply src/components/panel/panelSections/panelRelationships.component.tsx:135]: please do this to remove the console error about things not having keys

File: src/lib/graphql/skylark/dynamicMutations.ts [hunk 10]
```diff
@@ -263,3 +267,70 @@ export const createUpdateObjectContentMutation = (
...
+    mutation: {
+      __name: `UPDATE_OBJECT_RELATIONSHIPS_${object.name}`,
+      __variables: {
+        uid: "String!",
+      },
+      updateRelationships: {
+        __aliasFor: object.operations.update.name,
+        __args: {
+          uid: new VariableType("uid"),
+          [object.name.toLowerCase()]: {
```
Comment (james-wallis) [inline src/lib/graphql/skylark/dynamicMutations.ts:322]: This should be `object.argName`

Comment (james-wallis) [reply src/lib/graphql/skylark/dynamicMutations.ts:322]: `object.operations.update.argName`

**Review by james-wallis — CHANGES_REQUESTED** [review]

File: src/lib/graphql/skylark/dynamicMutations.ts (hunk 10 above)
Comment (freixedelo) [reply src/lib/graphql/skylark/dynamicMutations.ts:322]: doesn't work apply this change

**Review by james-wallis — APPROVED** [review]
Nice work!

**Pull request merged**